domain/upgrade/state/state.go

```diff
@@ -375,53 +349,98 @@ AND (
+// error if there is no active upgrade.
 func (st *State) ActiveUpgrade(ctx context.Context) (domainupgrade.UUID, error) {
 	db, err := st.DB()
 	if err != nil {
 		return "", errors.Trace(err)
 	}
-	var activeUpgrade domainupgrade.UUID
-	q := "SELECT (uuid) FROM upgrade_info WHERE state_type_id < ?"
 
-	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
-		row := tx.QueryRowContext(ctx, q, upgrade.StepsCompleted)
-		if err := row.Scan(&activeUpgrade); err != nil {
+	info := Info{
+		StateIDType: int(upgrade.StepsCompleted),
+	}
+
+	stmt, err := st.Prepare(`
+SELECT &Info.uuid
+FROM upgrade_info 
+WHERE state_type_id < $Info.state_type_id
+`, info)
+	if err != nil {
+		return "", errors.Annotate(err, "preparing select active upgrade statement")
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, info).Get(&info)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Annotatef(upgradeerrors.NotFound, "active upgrade")
+		} else if err != nil {
 			return errors.Trace(err)
 		}
 		return nil
 	})
-	return activeUpgrade, errors.Trace(err)
+	return domainupgrade.UUID(info.UUID), errors.Trace(domain.CoerceError(err))
 }
 
-// UpgradeInfo returns the upgrade info for the provided upgradeUUID
+// UpgradeInfo returns the upgrade info for the provided upgradeUUID. It returns
+// a NotFound error if the upgrade does not exist.
 func (st *State) UpgradeInfo(ctx context.Context, upgradeUUID domainupgrade.UUID) (upgrade.Info, error) {
 	db, err := st.DB()
 	if err != nil {
 		return upgrade.Info{}, errors.Trace(err)
 	}
+	info := Info{
+		UUID: upgradeUUID.String(),
+	}
 
-	q := `
-SELECT uuid, previous_version, target_version, upgrade_state_type.id 
+	stmt, err := st.Prepare(`
+SELECT &Info.*
 FROM upgrade_info 
-	LEFT JOIN upgrade_state_type
-		ON upgrade_info.state_type_id = upgrade_state_type.id
-WHERE uuid = ?
-	`
-
-	var upgradeInfoRow info
-	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
-		row := tx.QueryRowContext(ctx, q, upgradeUUID)
-		if err := row.Scan(&upgradeInfoRow.UUID, &upgradeInfoRow.PreviousVersion, &upgradeInfoRow.TargetVersion, &upgradeInfoRow.State); err != nil {
-			return errors.Trace(err)
+WHERE uuid = $Info.uuid
+`, info)
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err = tx.Query(ctx, stmt, info).Get(&info)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Annotatef(upgradeerrors.NotFound, "upgrade %q", upgradeUUID)
 		}
-		return nil
+		return err
 	})
 	if err != nil {
-		return upgrade.Info{}, errors.Trace(err)
+		return upgrade.Info{}, errors.Trace(domain.CoerceError(err))
 	}
 
-	result, err := upgradeInfoRow.ToUpgradeInfo()
+	result, err := info.ToUpgradeInfo()
 	return result, errors.Trace(err)
 }
+
+// updateState updates the state of an ongoing upgrade.
+func (st *State) updateState(ctx context.Context, tx *sqlair.TX, uuid string, from upgrade.State, to upgrade.State) error {
```

Comment:
Maybe should should check the transition `from` -> `to` is a valid transition?

Reply:
Good idea